### ISSUE TITLE
DI Misconceptions section in documentation seems misleading

### ISSUE THREAD

**Issue by schweeneh** [ssannandeji/Zenject-2019#561]
First off, thanks for creating this DI framework, Unity was sorely needing something like this.

This concerns the documentation section about common DI misconceptions. You write:

> A good rule of thumb instead is to only create interfaces when the class has more than one implementation (this is known, by the way, as the Reused Abstraction Principle)

While I agree that it is generally true that not all classes need interfaces, I think this is a very small subset of classes and when in doubt, extract an interface. This is because as soon as you use the new keyword inside a class method, you've tightly coupled your code and you lose all benefits of DI, regardless of whether or not the class has multiple implementations. I would even go so far as to say that the majority of classes don't have multiple implementations. For example, would you really advise people not to create a repository interface for a database adapter? What about an interface for a class that accesses the file system? With these two examples, if you don't create an interface and use DI, you can't write unit tests for any of the client code that relies on these services.

This part of the documentation threw me for a second, and I think it could be a big set back for someone learning DI for the first time.

Thanks for listening.

**Comment by cooloon** [comment]
If unit test is planned, you need to create interfaces for those classes. Otherwise you aren't gonna need it.

I think Zenject has three benefits:
1. Easy to test(as you said)
2. Easy to make Loosely-Coupled classes
3. Easy to control instance lifecycle

Without interfaces, benefit 2 and 3 are still available.

**Comment by cooloon** [comment]
test

**Comment by schweeneh** [comment]
That's true, 2 and 3 are still available, but I would still argue that it's a not a misconception to use interfaces for types that don't have multiple implementations. There's a very good reason for creating interfaces when a type only has one implementation, and that's testing. I don't consider unit tests optional, but I supposed if you do, then what you're saying is valid. It just seems like you're guiding people away from a totally harmless pattern, which is providing interfaces for every injected service. 

There's also another benefit to using interfaces for single implementations and that's the ease of replacing the component. It's a lesser benefit, but it's still valid.

**Comment by cooloon** [comment]
You may have a point. Unite tests are vital for harmless world.

Reading your first comment at the top again, I found your misunderstanding. Currently you have two options:

1. Use interfaces and enjoy all benefits of DI.
2. Avoid interfaces and use `new` keyword. This loses all benefits of DI.

But you can legally bind any classes without interfaces. If you have a single-implemented class, just bind that type.
```csharp
Container.Bind<SingleImplementedClass>().AsSingle().NonLazy();
```
By writing this, an instance of SingleImplementedClass will be created and injected automatically.

P.S. I'm not the author of Zenject or the README document. My comments are just my opinion☺️

**Comment by schweeneh** [comment]
Yes, you're right about binding a type to itself, that works for classes without interfaces. The thing I said about the new keyword is not a valid complaint. I still would hope that the Zenject authors consider changing the README, I still find the strong statement about that particular DI misconception to be potentially misleading for folks new to DI.

**Comment by cooloon** [comment]
What kind of single-implemented classes do you create interfaces for? It's reasonable to create interfaces for DatabaseAdapter, FileSystem, or NetworkManager. But, say, do you create interfaces for CsvParser or HitPointView in practice?
